{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Examples using Biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "random_seq = \"MKFTKPDRTCEVELFTAAFGLLRWAMQFINIPEAHGEPRLDILGKDEVQPQTQRWRRNIAECKFHVWKSPIVSRCQCKDFLGARFALKKSHILAQLVALELKVPTAAKKRQPVLLESPHFILDNCGTDVVGIIVAIVIDLPVVVAQLAKFKCDTCTDGSESWGGLITMYLVKPGLVKDPLKIRSDYEEVILTAGYDGVQTFSDASSGFDGLEHKYLIENVDNNDFLRAPIMLRLIRKYGGPVLQMLIRPSFLAAGAAKGKLIGQERMIRGRCLEKDFIWSNDGVHSLDCAFCRVLRFGMRYAWNGLMVEIERRLISITDLDMNIQSEKPERKYDLGHISRMNLSPFDGAKGKNLWTRDDAEFKLAQYNDADFSAREGRVSLLSEDKGDSPKLAQRRNEMGNSIAICDENACRLLHMAYLTEKEGTNAKDDVPKVSKGDPPEKHMTNEFKALLVLQLVNDYISGCLVDAKIPIPDVGQFPHFIKMPNKAAGSTLREVHIKI\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_sequence = Seq(random_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "M 14\n",
      "K 38\n",
      "F 22\n",
      "T 17\n",
      "K 38\n",
      "P 24\n",
      "D 36\n",
      "R 31\n",
      "T 17\n",
      "C 13\n",
      "E 29\n",
      "V 31\n",
      "E 29\n",
      "L 52\n",
      "F 22\n",
      "T 17\n",
      "A 37\n",
      "A 37\n",
      "F 22\n",
      "G 34\n",
      "L 52\n",
      "L 52\n",
      "R 31\n",
      "W 7\n",
      "A 37\n",
      "M 14\n",
      "Q 16\n",
      "F 22\n",
      "I 34\n",
      "N 19\n",
      "I 34\n",
      "P 24\n",
      "E 29\n",
      "A 37\n",
      "H 11\n",
      "G 34\n",
      "E 29\n",
      "P 24\n",
      "R 31\n",
      "L 52\n",
      "D 36\n",
      "I 34\n",
      "L 52\n",
      "G 34\n",
      "K 38\n",
      "D 36\n",
      "E 29\n",
      "V 31\n",
      "Q 16\n",
      "P 24\n",
      "Q 16\n",
      "T 17\n",
      "Q 16\n",
      "R 31\n",
      "W 7\n",
      "R 31\n",
      "R 31\n",
      "N 19\n",
      "I 34\n",
      "A 37\n",
      "E 29\n",
      "C 13\n",
      "K 38\n",
      "F 22\n",
      "H 11\n",
      "V 31\n",
      "W 7\n",
      "K 38\n",
      "S 25\n",
      "P 24\n",
      "I 34\n",
      "V 31\n",
      "S 25\n",
      "R 31\n",
      "C 13\n",
      "Q 16\n",
      "C 13\n",
      "K 38\n",
      "D 36\n",
      "F 22\n",
      "L 52\n",
      "G 34\n",
      "A 37\n",
      "R 31\n",
      "F 22\n",
      "A 37\n",
      "L 52\n",
      "K 38\n",
      "K 38\n",
      "S 25\n",
      "H 11\n",
      "I 34\n",
      "L 52\n",
      "A 37\n",
      "Q 16\n",
      "L 52\n",
      "V 31\n",
      "A 37\n",
      "L 52\n",
      "E 29\n",
      "L 52\n",
      "K 38\n",
      "V 31\n",
      "P 24\n",
      "T 17\n",
      "A 37\n",
      "A 37\n",
      "K 38\n",
      "K 38\n",
      "R 31\n",
      "Q 16\n",
      "P 24\n",
      "V 31\n",
      "L 52\n",
      "L 52\n",
      "E 29\n",
      "S 25\n",
      "P 24\n",
      "H 11\n",
      "F 22\n",
      "I 34\n",
      "L 52\n",
      "D 36\n",
      "N 19\n",
      "C 13\n",
      "G 34\n",
      "T 17\n",
      "D 36\n",
      "V 31\n",
      "V 31\n",
      "G 34\n",
      "I 34\n",
      "I 34\n",
      "V 31\n",
      "A 37\n",
      "I 34\n",
      "V 31\n",
      "I 34\n",
      "D 36\n",
      "L 52\n",
      "P 24\n",
      "V 31\n",
      "V 31\n",
      "V 31\n",
      "A 37\n",
      "Q 16\n",
      "L 52\n",
      "A 37\n",
      "K 38\n",
      "F 22\n",
      "K 38\n",
      "C 13\n",
      "D 36\n",
      "T 17\n",
      "C 13\n",
      "T 17\n",
      "D 36\n",
      "G 34\n",
      "S 25\n",
      "E 29\n",
      "S 25\n",
      "W 7\n",
      "G 34\n",
      "G 34\n",
      "L 52\n",
      "I 34\n",
      "T 17\n",
      "M 14\n",
      "Y 10\n",
      "L 52\n",
      "V 31\n",
      "K 38\n",
      "P 24\n",
      "G 34\n",
      "L 52\n",
      "V 31\n",
      "K 38\n",
      "D 36\n",
      "P 24\n",
      "L 52\n",
      "K 38\n",
      "I 34\n",
      "R 31\n",
      "S 25\n",
      "D 36\n",
      "Y 10\n",
      "E 29\n",
      "E 29\n",
      "V 31\n",
      "I 34\n",
      "L 52\n",
      "T 17\n",
      "A 37\n",
      "G 34\n",
      "Y 10\n",
      "D 36\n",
      "G 34\n",
      "V 31\n",
      "Q 16\n",
      "T 17\n",
      "F 22\n",
      "S 25\n",
      "D 36\n",
      "A 37\n",
      "S 25\n",
      "S 25\n",
      "G 34\n",
      "F 22\n",
      "D 36\n",
      "G 34\n",
      "L 52\n",
      "E 29\n",
      "H 11\n",
      "K 38\n",
      "Y 10\n",
      "L 52\n",
      "I 34\n",
      "E 29\n",
      "N 19\n",
      "V 31\n",
      "D 36\n",
      "N 19\n",
      "N 19\n",
      "D 36\n",
      "F 22\n",
      "L 52\n",
      "R 31\n",
      "A 37\n",
      "P 24\n",
      "I 34\n",
      "M 14\n",
      "L 52\n",
      "R 31\n",
      "L 52\n",
      "I 34\n",
      "R 31\n",
      "K 38\n",
      "Y 10\n",
      "G 34\n",
      "G 34\n",
      "P 24\n",
      "V 31\n",
      "L 52\n",
      "Q 16\n",
      "M 14\n",
      "L 52\n",
      "I 34\n",
      "R 31\n",
      "P 24\n",
      "S 25\n",
      "F 22\n",
      "L 52\n",
      "A 37\n",
      "A 37\n",
      "G 34\n",
      "A 37\n",
      "A 37\n",
      "K 38\n",
      "G 34\n",
      "K 38\n",
      "L 52\n",
      "I 34\n",
      "G 34\n",
      "Q 16\n",
      "E 29\n",
      "R 31\n",
      "M 14\n",
      "I 34\n",
      "R 31\n",
      "G 34\n",
      "R 31\n",
      "C 13\n",
      "L 52\n",
      "E 29\n",
      "K 38\n",
      "D 36\n",
      "F 22\n",
      "I 34\n",
      "W 7\n",
      "S 25\n",
      "N 19\n",
      "D 36\n",
      "G 34\n",
      "V 31\n",
      "H 11\n",
      "S 25\n",
      "L 52\n",
      "D 36\n",
      "C 13\n",
      "A 37\n",
      "F 22\n",
      "C 13\n",
      "R 31\n",
      "V 31\n",
      "L 52\n",
      "R 31\n",
      "F 22\n",
      "G 34\n",
      "M 14\n",
      "R 31\n",
      "Y 10\n",
      "A 37\n",
      "W 7\n",
      "N 19\n",
      "G 34\n",
      "L 52\n",
      "M 14\n",
      "V 31\n",
      "E 29\n",
      "I 34\n",
      "E 29\n",
      "R 31\n",
      "R 31\n",
      "L 52\n",
      "I 34\n",
      "S 25\n",
      "I 34\n",
      "T 17\n",
      "D 36\n",
      "L 52\n",
      "D 36\n",
      "M 14\n",
      "N 19\n",
      "I 34\n",
      "Q 16\n",
      "S 25\n",
      "E 29\n",
      "K 38\n",
      "P 24\n",
      "E 29\n",
      "R 31\n",
      "K 38\n",
      "Y 10\n",
      "D 36\n",
      "L 52\n",
      "G 34\n",
      "H 11\n",
      "I 34\n",
      "S 25\n",
      "R 31\n",
      "M 14\n",
      "N 19\n",
      "L 52\n",
      "S 25\n",
      "P 24\n",
      "F 22\n",
      "D 36\n",
      "G 34\n",
      "A 37\n",
      "K 38\n",
      "G 34\n",
      "K 38\n",
      "N 19\n",
      "L 52\n",
      "W 7\n",
      "T 17\n",
      "R 31\n",
      "D 36\n",
      "D 36\n",
      "A 37\n",
      "E 29\n",
      "F 22\n",
      "K 38\n",
      "L 52\n",
      "A 37\n",
      "Q 16\n",
      "Y 10\n",
      "N 19\n",
      "D 36\n",
      "A 37\n",
      "D 36\n",
      "F 22\n",
      "S 25\n",
      "A 37\n",
      "R 31\n",
      "E 29\n",
      "G 34\n",
      "R 31\n",
      "V 31\n",
      "S 25\n",
      "L 52\n",
      "L 52\n",
      "S 25\n",
      "E 29\n",
      "D 36\n",
      "K 38\n",
      "G 34\n",
      "D 36\n",
      "S 25\n",
      "P 24\n",
      "K 38\n",
      "L 52\n",
      "A 37\n",
      "Q 16\n",
      "R 31\n",
      "R 31\n",
      "N 19\n",
      "E 29\n",
      "M 14\n",
      "G 34\n",
      "N 19\n",
      "S 25\n",
      "I 34\n",
      "A 37\n",
      "I 34\n",
      "C 13\n",
      "D 36\n",
      "E 29\n",
      "N 19\n",
      "A 37\n",
      "C 13\n",
      "R 31\n",
      "L 52\n",
      "L 52\n",
      "H 11\n",
      "M 14\n",
      "A 37\n",
      "Y 10\n",
      "L 52\n",
      "T 17\n",
      "E 29\n",
      "K 38\n",
      "E 29\n",
      "G 34\n",
      "T 17\n",
      "N 19\n",
      "A 37\n",
      "K 38\n",
      "D 36\n",
      "D 36\n",
      "V 31\n",
      "P 24\n",
      "K 38\n",
      "V 31\n",
      "S 25\n",
      "K 38\n",
      "G 34\n",
      "D 36\n",
      "P 24\n",
      "P 24\n",
      "E 29\n",
      "K 38\n",
      "H 11\n",
      "M 14\n",
      "T 17\n",
      "N 19\n",
      "E 29\n",
      "F 22\n",
      "K 38\n",
      "A 37\n",
      "L 52\n",
      "L 52\n",
      "V 31\n",
      "L 52\n",
      "Q 16\n",
      "L 52\n",
      "V 31\n",
      "N 19\n",
      "D 36\n",
      "Y 10\n",
      "I 34\n",
      "S 25\n",
      "G 34\n",
      "C 13\n",
      "L 52\n",
      "V 31\n",
      "D 36\n",
      "A 37\n",
      "K 38\n",
      "I 34\n",
      "P 24\n",
      "I 34\n",
      "P 24\n",
      "D 36\n",
      "V 31\n",
      "G 34\n",
      "Q 16\n",
      "F 22\n",
      "P 24\n",
      "H 11\n",
      "F 22\n",
      "I 34\n",
      "K 38\n",
      "M 14\n",
      "P 24\n",
      "N 19\n",
      "K 38\n",
      "A 37\n",
      "A 37\n",
      "G 34\n",
      "S 25\n",
      "T 17\n",
      "L 52\n",
      "R 31\n",
      "E 29\n",
      "V 31\n",
      "H 11\n",
      "I 34\n",
      "K 38\n",
      "I 34\n"
     ]
    }
   ],
   "source": [
    "for i in range(0, len(protein_sequence)):\n",
    "\t# print the count of each base\n",
    "\tprint(protein_sequence[i], protein_sequence.count(protein_sequence[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GC content (%)\n",
    "\n",
    "def get_gc_content(protein_sequence: str) -> float:\n",
    "    \"\"\"Calculate the GC content percentage of a protein sequence.\"\"\"\n",
    "    gc_count = protein_sequence.count(\"G\") + protein_sequence.count(\"C\")\n",
    "    return (gc_count / len(protein_sequence)) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_overlap(protein_sequence: str, top_bound: int) -> dict:\n",
    "    \"\"\"Find the overlap of sequences with window sizes from 1 to top_bound.\"\"\"\n",
    "    overlap_counts = {}\n",
    "    for window_size in range(1, top_bound + 1):\n",
    "        for i in range(len(protein_sequence) - window_size + 1):\n",
    "            window = protein_sequence[i:i + window_size]\n",
    "            if window in overlap_counts:\n",
    "                overlap_counts[window] += 1\n",
    "            else:\n",
    "                overlap_counts[window] = 1\n",
    "    return overlap_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_amino_acid_composition(protein_sequence: str) -> dict:\n",
    "    \"\"\"Calculate the counts of each amino acid in the protein sequence.\"\"\"\n",
    "    protein_analysis = ProteinAnalysis(protein_sequence)\n",
    "    amino_acid_composition = protein_analysis.count_amino_acids()\n",
    "    return amino_acid_composition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_motif_presence(protein_sequence: str, motif: str) -> int:\n",
    "\t\"\"\"Check the presence of a specific motif in the protein sequence.\"\"\"\n",
    "\treturn 1 if motif in protein_sequence else 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_analysis_features = [\"molecular_weight\", \"aromaticity\", \"instability_index\", \"isoelectric_point\"]\n",
    "\n",
    "def add_protein_analysis_features(protein_sequence, protein_analysis_features):\n",
    "\tprotein_analysis = ProteinAnalysis(protein_sequence)\n",
    "\tdata = {}\n",
    "\tfor feature_name in protein_analysis_features:\n",
    "\t\tfunction = getattr(protein_analysis, feature_name)\n",
    "\t\tresult = function()\n",
    "\t\ttry:\n",
    "\t\t\tdata[feature_name] = float(result)\n",
    "\t\texcept ValueError:\n",
    "\t\t\tdata[feature_name] = result\n",
    "\treturn data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate GC content\n",
    "gc_content = get_gc_content(protein_sequence)\n",
    "\n",
    "# Find overlapping sequences\n",
    "top_bound = 3\n",
    "overlap_counts = find_overlap(protein_sequence, top_bound)\n",
    "\n",
    "data = {\"GC_content\": gc_content}\n",
    "for key, value in overlap_counts.items():\n",
    "    data[key] = value\n",
    "\n",
    "# Calculate amino acid composition\n",
    "amino_acid_composition = get_amino_acid_composition(protein_sequence)\n",
    "\n",
    "# Create DataFrame with protein sequence and features\n",
    "df = pd.DataFrame(data, index=[0])\n",
    "df[\"protein_sequence\"] = str(protein_sequence)\n",
    "df[\"protein_length\"] = len(str(protein_sequence))\n",
    "\n",
    "# Add features for each amino acid count\n",
    "for amino_acid, count in amino_acid_composition.items():\n",
    "    feature_name = f\"count_{amino_acid}\"\n",
    "    df[feature_name] = count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "# motifs\n",
    "\n",
    "# Check presence of motifs or functional domains\n",
    "motif1 = \"RCE\"\n",
    "motif2 = \"LKA\"\n",
    "motif3 = \"GKDE\"\n",
    "motif4 = \"KPTAAK\"\n",
    "\n",
    "for motif in [motif1, motif2, motif3, motif4]:\n",
    "\tdf[f\"contains_{motif}\"] = check_motif_presence(protein_sequence, motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'molecular_weight': 56295.65200000005, 'aromaticity': 0.078, 'instability_index': 43.242200000000004, 'isoelectric_point': 8.299804878234863}\n"
     ]
    }
   ],
   "source": [
    "# features for data protein analysis features\n",
    "\n",
    "features_data_protein_analysis_features = add_protein_analysis_features(protein_sequence, protein_analysis_features)\n",
    "\n",
    "print(features_data_protein_analysis_features)\n",
    "\n",
    "for feature_name in protein_analysis_features:\n",
    "    df[feature_name] = pd.to_numeric(features_data_protein_analysis_features[feature_name], errors='coerce')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GC_content</th>\n",
       "      <th>(M)</th>\n",
       "      <th>(K)</th>\n",
       "      <th>(F)</th>\n",
       "      <th>(T)</th>\n",
       "      <th>(P)</th>\n",
       "      <th>(D)</th>\n",
       "      <th>(R)</th>\n",
       "      <th>(C)</th>\n",
       "      <th>(E)</th>\n",
       "      <th>...</th>\n",
       "      <th>count_W</th>\n",
       "      <th>count_Y</th>\n",
       "      <th>contains_RCE</th>\n",
       "      <th>contains_LKA</th>\n",
       "      <th>contains_GKDE</th>\n",
       "      <th>contains_KPTAAK</th>\n",
       "      <th>molecular_weight</th>\n",
       "      <th>aromaticity</th>\n",
       "      <th>instability_index</th>\n",
       "      <th>isoelectric_point</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9.4</td>\n",
       "      <td>14</td>\n",
       "      <td>38</td>\n",
       "      <td>22</td>\n",
       "      <td>17</td>\n",
       "      <td>24</td>\n",
       "      <td>36</td>\n",
       "      <td>31</td>\n",
       "      <td>13</td>\n",
       "      <td>29</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>56295.652</td>\n",
       "      <td>0.078</td>\n",
       "      <td>43.2422</td>\n",
       "      <td>8.299805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 791 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   GC_content  (M)  (K)  (F)  (T)  (P)  (D)  (R)  (C)  (E)  ...  count_W  \\\n",
       "0         9.4   14   38   22   17   24   36   31   13   29  ...        7   \n",
       "\n",
       "   count_Y  contains_RCE  contains_LKA  contains_GKDE  contains_KPTAAK  \\\n",
       "0       10             0             0              1                0   \n",
       "\n",
       "   molecular_weight  aromaticity  instability_index  isoelectric_point  \n",
       "0         56295.652        0.078            43.2422           8.299805  \n",
       "\n",
       "[1 rows x 791 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to csv file\n",
    "\n",
    "df.to_csv('protein_data.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biopython-examples-ZQ7hZzro",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
